packages/flutter_tools/lib/src/base/dds.dart

```diff
@@ -4,112 +4,329 @@
 
 import 'dart:async';
 
-import 'package:dds/dds.dart' as dds;
 import 'package:meta/meta.dart';
 
-import 'common.dart';
-import 'context.dart';
+import '../convert.dart';
+import '../device.dart';
 import 'io.dart' as io;
 import 'logger.dart';
+import 'platform.dart' as p;
 
-// TODO(fujino): This should be direct injected, rather than mutable global state.
-@visibleForTesting
-Future<dds.DartDevelopmentService> Function(
+/// A representation of the current DDS state including:
+///
+/// - The process the external DDS instance is running in
+/// - The service URI DDS is being served on
+/// - The URI DevTools is being served on, if applicable
+/// - The URI DTD is being served on, if applicable
+typedef DartDevelopmentServiceInstance = ({
+  io.Process? process,
+  Uri? serviceUri,
+  Uri? devToolsUri,
+  Uri? dtdUri,
+});
+
+/// The default DDSLauncherCallback used to spawn DDS.
+Future<DartDevelopmentServiceInstance> defaultStartDartDevelopmentService(
   Uri remoteVmServiceUri, {
-  bool enableAuthCodes,
-  bool ipv6,
+  required bool enableAuthCodes,
+  required bool ipv6,
+  required bool enableDevTools,
+  required List<String> cachedUserTags,
   Uri? serviceUri,
-  List<String> cachedUserTags,
-  dds.UriConverter? uriConverter,
-}) ddsLauncherCallback = dds.DartDevelopmentService.startDartDevelopmentService;
+  String? google3WorkspaceRoot,
+  Uri? devToolsServerAddress,
+}) async {
+  final String exe = const p.LocalPlatform().executable;
+  final io.Process process = await io.Process.start(
+    exe,
+    <String>[
+      'development-service',
+      '--vm-service-uri=$remoteVmServiceUri',
+      if (serviceUri != null) ...<String>[
+        '--bind-address=${serviceUri.host}',
+        '--bind-port=${serviceUri.port}',
+      ],
+      if (!enableAuthCodes) '--disable-service-auth-codes',
+      // TODO(bkonyi): uncomment when ready to serve DevTools from DDS.
+      // if (enableDevTools) '--serve-devtools',
+      if (google3WorkspaceRoot != null)
+        '--google3-workspace-root=$google3WorkspaceRoot',
+      // TODO(bkonyi): uncomment when ready to serve DevTools from DDS.
+      // if (devToolsServerAddress != null)
+      //  '--devtools-server-address=$devToolsServerAddress',
+      for (final String tag in cachedUserTags) '--cached-user-tags=$tag',
+    ],
+  );
+  final Completer<DartDevelopmentServiceInstance> completer =
+      Completer<DartDevelopmentServiceInstance>();
+  late StreamSubscription<String> stderrSub;
+  stderrSub = process.stderr.transform(utf8.decoder).listen((String event) {
+    final Map<String, dynamic> result =
+        json.decode(event) as Map<String, dynamic>;
```

Comment:
I chatted with bquinlan, and we agreed that this is not safe--decoding JSON *may* fail if the first event emitted by the stderr stream does not include ALL the bytes of the JSON object.

If the assumption is that this is the only thing `dart development-service` will output to STDERR is this JSON blob, can it then close its stderr pipe? Then from the tool, we could call:

```dart
  final String event = await process.stderr.transform(utf8.decoder).join();
  final Map<String, Object?> result = json.decode(event);
```

If it's not feasible to close stderr after sending the JSON from the other side, then we should catch json decoder errors and in that case, assume we only have part of the message, and store what we have in a variable in the higher scope, and then concatenate the next time the `.listen()` callback fires.

Reply:
Ah, good point. We've been using similar logic in the standalone VM for years and haven't encountered any reported issues, but it's better to be safe than sorry.

I think we can achieve this using the `JsonDecoder` as a transformer, which will handle the chunking automatically.

Reply:
Ah, `json.decoder` waits for the underlying stream to close before processing the event... closing `stderr` on the DDS side needs to be done as well.

Reply:
Nice, thanks!